x-pack/plugins/security/server/lib/authorization/saved_objects/ensure_saved_objects_privileges.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { get } from 'lodash';
+import { Namespace } from 'src/legacy/server/saved_objects';
+import { DEFAULT_SPACE_ID } from '../../../../../spaces/common/constants';
+import { CheckPrivileges, CheckPrivilegesAtResourceResponse } from '../check_privileges';
+import { Actions } from '../actions';
+
+export type SavedObjectsOperation =
+  | 'create'
+  | 'update'
+  | 'delete'
+  | 'get'
+  | 'bulk_get'
+  | 'bulk_create'
+  | 'find';
+
+interface Deps {
+  errors: {
+    decorateGeneralError: (error: Error, reason: string) => Error;
+    decorateForbiddenError: (error: Error) => Error;
+  };
+  spacesEnabled: boolean;
+  checkPrivileges: CheckPrivileges;
+  actionsService: Actions;
+  auditLogger: any;
+}
+
+export type EnsureSavedObjectsPrivileges = (
+  typeOrTypes: string | string[] | undefined,
+  operation: SavedObjectsOperation,
+  namespace: Namespace | undefined,
+  args: any
+) => Promise<void>;
+
+export function ensureSavedObjectsPrivilegesFactory(deps: Deps) {
+  const { errors, spacesEnabled, actionsService, auditLogger, checkPrivileges } = deps;
+
+  const ensureSavedObjectsPrivileges: EnsureSavedObjectsPrivileges = async (
+    typeOrTypes: string | string[] | undefined,
+    operation: SavedObjectsOperation,
+    namespace: Namespace | undefined,
+    args: any
+  ) => {
+    const types = normalizeTypes(typeOrTypes);
+    const actionsToTypesMap = new Map(
+      types.map(type => [actionsService.savedObject.get(type, operation), type] as [string, string])
+    );
+    const actions = Array.from(actionsToTypesMap.keys());
+
+    let privilegeResponse: CheckPrivilegesAtResourceResponse;
+
+    try {
+      if (spacesEnabled) {
```

Comment:
Instead of using the `spacesEnabled` boolean, we could theoretically use `namespace` being undefined to determine that spaces isn't enabled and that we should be checking "globally". I don't know how I feel about this, what do you think?

Reply:
I'm torn on this. I like that it's more direct, and requires one less dependency, but it also feels a bit confusing that the existence of a namespace means that privileges are checked at a space, as opposed to globally. 

What happens if another wrapper suddenly provides an empty namespace when we weren't expecting one (assuming spaces is disabled)? A test would catch this scenario, but I don't expect functional tests to run with spaces disabled, unless they're written by our team.

Reply:
> What happens if another wrapper suddenly provides an empty namespace when we weren't expecting one (assuming spaces is disabled)? A test would catch this scenario, but I don't expect functional tests to run with spaces disabled, unless they're written by our team.

Good point. Let's just keep it the way it is then...